### PR TITLE
Validate AZ configuration correctly 

### LINKED ISSUE #4891: `availabilityZones` parsing broken after upgrading to eksctl 0.86.0
**What were you trying to accomplish?**
Create a new cluster with `eksctl 0.86.0`

**What happened?**
Defined `availabilityZones` in ClusterConfig but it randomly generated subnets but not honor the YAML setting.

**How to reproduce it?**
Create a minimal ClusterConfig (see blow) then execute the command as `eksctl create cluster -f ./experimental.yaml --dry-run` and you should see bouncing `availabilityZones` being generated. Tried to search root cause from commit history, seems like it's related to #4735 and #4837.

Additional note, `eksctl create cluster --nodes=us-east-1a,us-east-1b ...` works as expected with no issue.

image

**Logs**

Testing input:
```
$ cat ./experimental.yaml
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: experimental
  region: us-east-1

availabilityZones:
- us-east-1a
- us-east-1b
```

Tool version:
```
$ eksctl version
0.86.0
```

Unexpected results:
```
 $ for i in $(seq 1 3); do (echo "[[ Round ${i} ]]";eksctl create cluster -f ./experimental.yaml --dry-run | grep 'availabilityZones' -A 2; echo); done
[[ Round 1 ]]
availabilityZones:
- us-east-1b
- us-east-1f

[[ Round 2 ]]
availabilityZones:
- us-east-1b
- us-east-1d

[[ Round 3 ]]
availabilityZones:
- us-east-1c
- us-east-1e
```

**Anything else we need to know?**

It's was worked as expected before `0.85.0`

Tool version:
```
$ ./eksctl-v0.85.0 version
0.85.0
```

Results:
```
$ for i in $(seq 1 3); do (echo "[[ Round ${i} ]]";./eksctl-v0.85.0 create cluster -f ./experimental.yaml --dry-run | grep 'availabilityZones' -A 2; echo); done
[[ Round 1 ]]
availabilityZones:
- us-east-1a
- us-east-1b

[[ Round 2 ]]
availabilityZones:
- us-east-1a
- us-east-1b

[[ Round 3 ]]
availabilityZones:
- us-east-1a
- us-east-1b
```

**Versions**

```
$ eksctl info
eksctl version: 0.86.0
kubectl version: v1.23.4
OS: darwin
```

### PR DESCRIPTION
**Pull request by nikimanoledaki**
### Description

Closes #4891

During some refactoring, the AZ validation that returned early if AZs were set was accidentally removed, so the AZ configuration was not respected. Reinstating this to fix that bug :) + adding unit-tests so that hopefully it doesn't happen again 😬 

### Checklist
- [x] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes
- [x] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [x] Backfilled missing tests for code in same general area :tada:
- [x] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Review by Skarlso — APPROVED** [review]

**Review by cPu1 — APPROVED** [review]

File: pkg/eks/api.go [hunk 1]
```diff
@@ -377,6 +377,13 @@ func (c *ClusterProvider) SetAvailabilityZones(spec *api.ClusterConfig, given []
 		return nil
 	}
 
+	if count := len(spec.AvailabilityZones); count != 0 {
+		if count < api.MinRequiredAvailabilityZones {
+			return api.ErrTooFewAvailabilityZones(spec.AvailabilityZones)
+		}
+		return nil
+	}
+
```
Comment (aclevername) [inline pkg/eks/api.go:386]: don't we already do this? https://github.com/weaveworks/eksctl/blob/df75054cf0b2696130f465ac7a0af7e4bbe7b851/pkg/apis/eksctl.io/v1alpha5/validation.go#L1296

Comment (Skarlso) [reply pkg/eks/api.go:386]: For given azs, but those don't come from the spec apparently.

Comment (Skarlso) [reply pkg/eks/api.go:386]: They come from `params`:

```go
		if err := ctl.SetAvailabilityZones(cfg, params.AvailabilityZones); err != nil {
			return err
		}
```

Comment (Skarlso) [reply pkg/eks/api.go:386]: Which is set by cli. But the extra check is for when it's coming from the yaml config. Pretty weird, could probably be refactored at some point.

Comment (aclevername) [reply pkg/eks/api.go:386]: > For given azs, but those don't come from the spec apparently.

I don't follow, both codes use `spec/cfg.AvailabilityZones`

Comment (nikimanoledaki) [reply pkg/eks/api.go:386]: Very good point but no it doesn't: this specific function is (only) used by `doCreateCluster`, which does not use `ValidateClusterConfig`. Moved this validation to `ValidateClusterConfig`, but the next step is to make all these `ctl` funcs actually use it. https://github.com/weaveworks/eksctl/issues/4862 addresses this :)

Comment (Skarlso) [reply pkg/eks/api.go:386]: See previous comment. :)

**Comment by nikimanoledaki** [comment]
Added unit tests (which required a slight refactor)

**Review by Skarlso — APPROVED** [review]
Very nice changes, well done Niki! :)

File: pkg/eks/api_test.go [hunk 2]
```diff
@@ -153,3 +156,97 @@ func mockDescribeImages(p *mockprovider.MockProvider, amiID string, matcher func
...
+			It("sets them as the AZs to be used", func() {
+				err := eks.SetAvailabilityZones(cfg, []string{"us-east-2a", "us-east-2b"}, provider.EC2(), "")
+				Expect(err).NotTo(HaveOccurred())
+			})
+		})
+
+		When("the given params contain too few AZs", func() {
+			It("returns an error", func() {
+				err := eks.SetAvailabilityZones(cfg, []string{"us-east-2a"}, provider.EC2(), "")
+				Expect(err).To(MatchError(ContainSubstring("only 1 zone(s) specified [us-east-2a], 2 are required (can be non-unique)")))
```
Comment (aclevername) [inline pkg/eks/api_test.go:182]: huge nit, here and a few other places isn't this an exact error match? So we could remove the contain substring part
```suggestion
				Expect(err).To(MatchError("only 1 zone(s) specified [us-east-2a], 2 are required (can be non-unique)"))
```

could totally be wrong but first glace looks like all the errors are internal, so its an exact error match

**Review by aclevername — APPROVED** [review]
1 nit

File: pkg/eks/api_test.go (hunk 2 above)
Comment (nikimanoledaki) [reply pkg/eks/api_test.go:182]: testing errors and jake - name a more dynamic duo, I'll wait

Comment (nikimanoledaki) [reply pkg/eks/api_test.go]: done!

**Pull request merged**